Wip/schemaorg automation2

rayslxu opened the pull request.

gcampax commented on tool/lib/webqa-manual-annotations.js:
```diff
@@ -254,7 +254,9 @@ const MANUAL_PROPERTY_CANONICAL_OVERRIDE = {
             'by', 'made by', 'written by', 'created by', 'authored by', 'uploaded by', 'submitted by'
-        ]
+        ],
+        reverse_verb: ['wrote', 'authored']
```
made? did? (here and elsewhere)

rayslxu replied:
I added "made" to creator field. We will need domain specific override in the future, otherwise some annotations will introduce ambiguity in some domains.

gcampax commented on tool/lib/webqa-manual-annotations.js:
```diff
@@ -369,22 +372,26 @@ const MANUAL_PROPERTY_CANONICAL_OVERRIDE = {
         property: ['#'],
-        passive_verb: ['played by', 'acted by']
+        passive_verb: ['played by', 'acted by'],
+        reverse_verb: ['acted', 'acted in']
```
What about "was in" and "is in" (as in "who was in titanic?") Is that something you can get with "preposition"? Is that a reversed verb?

rayslxu replied:
Added (it should be reverse verb. preposition would be "in Leonardo DiCapiro" not "in Titanic")

gcampax commented on languages/thingtalk/load-thingpedia.js:
```diff
@@ -151,6 +152,12 @@ class ThingpediaLoader {
         this._grammar.addRule('out_param_' + typestr, [canonical], this._runtime.simpleCombine(() => new Ast.Value.VarRef(pname)));
 
+        if (!pname.startsWith('address') && !pname.endsWith('Count') && !pname.startsWith('numberOf')) {
+            let postags = posTag(canonical.split(' '));
```
No. The POS tag should be in the annotation name already.

rayslxu replied:
Fixed in d0c94f3. The reason to have this is we always generate base form even if there is no noun phrase found. But I think it's fine to remove it - it will generate werid sentence but won't generate ambiguity.

gcampax commented on lib/i18n/american-english.js:
```diff
@@ -377,6 +377,16 @@ function isGoodSentence(sentence) {
     return true;
 }
 
+function isZipcode(word) {
```
Are you going to care about British postcodes? The format is 6 characters: 3 numbers and 3 letters, alternated, often displayed as 2 words of 3 characters

rayslxu replied:
probably in british-english.js?

gcampax replied:
I guess? I'm still passing en-US as the locale for MultiWOZ, I suppose I should pass en-GB...

gcampax replied:
I'm also not sure we'll copy over all code from american-english into british-english, and then canadian-english, australian-english, south-african-english, nigerian-english... I think we'll have to revisit the language pack concept to pass the locale as a parameter, for the truly country specific stuff (phone number intl prefixes are another case like this)

rayslxu commented on tool/lib/file_parameter_provider.js:
```diff
@@ -41,7 +41,14 @@ module.exports = class FileParameterProvider {
+            if (row.length === 4) {
+                [stringOrEntity, locale, type, filepath] = row;
+            } else {
+                locale = 'en-US';
```
Agree, the commit is reverted.

rayslxu commented on lib/i18n/american-english.js:
```diff
@@ -333,9 +353,28 @@ function isGoodWord(word) {
+    // filter out sentences containing only numbers
+    let allNumber = true;
+    for (let word of words) {
+        if (isNaN(word))
```
Replaced with a regex in c2121f

rayslxu commented on lib/i18n/american-english.js:
```diff
@@ -14,6 +14,26 @@ const { coin } = require('../random');
+    'm', 'o', 're', 've', 'y', 'ain', 'aren', "aren't", 'couldn', "couldn't", 'didn', "didn't", 'doesn', "doesn't",
+    'hadn', "hadn't", 'hasn', "hasn't", 'haven', "haven't", 'isn', "isn't", 'ma', 'mightn', "mightn't", 'mustn',
+    "mustn't", 'needn', "needn't", 'shan', "shan't", 'shouldn', "shouldn't", 'wasn', "wasn't", 'weren', "weren't",
+    'won', "won't", 'wouldn', "wouldn't"];
```
Fixed

rayslxu commented on lib/utils.js:
```diff
@@ -237,12 +238,44 @@ function makeDummyEntity(token) {
         return ENTITIES[entityType](Number(entityIndex));
 }
 
+function sampleString(words, langPack, rng=Math.random) {
```
Fixed

rayslxu commented on tool/webqa-normalize-data.js:
```diff
@@ -445,11 +445,37 @@ class Normalizer {
         }
     }
 
+    _preprocess(input) {
```
Added comment for this one in 347e2f6. We will need another round of jsdoc, but later i think...

rayslxu commented on tool/lib/webqa-canonical-annotator.js:
```diff
@@ -95,10 +95,14 @@ class AutoCanonicalAnnotator {
+
+                    // if it's the only people entity, adding adjective form
+                    if (typestr.endsWith(':Person'))
+                        arg.metadata.canonical.adjective = ["# 's", '#'];
```
Added in 347e2f6.

rayslxu commented on languages/thingtalk/en/commands.genie:
```diff
@@ -401,6 +401,18 @@ projection_question = {
         return C.sayProjection(proj);
     };
+    'who' proj:who_rv_projection_table '?' => {
+        // FIXME: this should not happen
```
I was just copying what we had for other projection... I added the assert in 505d353

rayslxu commented on languages/thingtalk/load-thingpedia.js:
```diff
@@ -20,6 +20,7 @@ const SchemaRetriever = ThingTalk.SchemaRetriever;
 const Units = ThingTalk.Units;
 
 const { clean, typeToStringSafe, makeFilter, makeAndFilter } = require('./utils');
+const { isHumanEntity } = require('../../lib/utils');
```
Fixed in d0c94f3

rayslxu commented on lib/utils.js:
```diff
@@ -25,6 +26,13 @@ class ValidationError extends Error {
+        .initial() // initial dictionary and pattern based tagging
+        .smooth() // further context based smoothing
+        .tags;
+}
```
Moved to `american-english.js` in 2c81990. Since schema.org is in English, the entire autoqa is heavily relying on english-specific heuristics. It will be another project to do other languages.

gcampax replied:
It's not just schema.org, right? It's any database schema with meaningful column names. Presumably, a depraved person could have a database schema with Italian column names... Although yeah English is probably the most important language.

rayslxu commented on tool/lib/webqa-canonical-annotator.js:
```diff
@@ -228,14 +229,31 @@ class AutoCanonicalAnnotator {
+                    //  to verb doesn't work, add the following heuristics
+                    const base = (new Inflectors(canonicals.base[0])).toSingular();
+                    if (base.endsWith('or') || base.endsWith('er'))
+                        canonicals.reverse_verb.push(base.slice(0, -2) + 'ed');
```
No, there isn't unfortunately....

gcampax replied:
It would probably be useful to add, and at the very list we should add all the words in schema.org as special cases.

rayslxu replied:
If we do this for all the words, isn't that contradict the concept of autoqa? we do have manual annotation for it.

gcampax replied:
It's not QA work, it's just natural language work. You could use WordNet or a dictionary or some other resource. You're already relying on people doing this work for plurals, for past tenses, and so on. There is no reason not to do this work, even for AutoQA.

rayslxu commented on tool/lib/webqa-template-extractor.js:
```diff
@@ -0,0 +1,218 @@
+
+    async _paraphrase(input, arg) {
+        // if debug file exists, use them directly
+        if (fs.existsSync(`./gpt2-paraphraser-out-${arg}.json`))
```
historical reasons obviously.... fixed in ab01603

rayslxu commented:
> Lots of comments, some pretty important to address.

I think I addressed most of the comments (except the one to turn noun to passive verb). The test rely on tokenizer, maybe we can merge this first and add test once we have your new tokenizer.
